relayer/src/link/relay_path.rs

```diff
@@ -693,103 +733,212 @@ impl<ChainA: ChainHandle, ChainB: ChainHandle> RelayPath<ChainA, ChainB> {
         self.recv_packet_acknowledged_on_src(&rp.packet)
     }
 
-    /// Returns `true` if the delay for this relaying path is zero.
-    /// Conversely, returns `false` if the delay is non-zero.
-    pub fn zero_delay(&self) -> bool {
-        self.channel.connection_delay == ZERO_DURATION
+    /// Returns the `processed_height` for the consensus state at specified height
+    fn update_height(
+        chain: &impl ChainHandle,
+        client_id: ClientId,
+        consensus_height: Height,
+    ) -> Result<Height, LinkError> {
+        let events = chain
+            .query_txs(QueryTxRequest::Client(QueryClientEventRequest {
+                height: Height::zero(),
+                event_id: WithBlockDataType::UpdateClient,
+                client_id,
+                consensus_height,
+            }))
+            .map_err(|e| LinkError::query(chain.id(), e))?;
+
+        // The handler may treat redundant updates as no-ops and emit `UpdateClient` events for them
+        // but the `processed_height` is the height at which the first `UpdateClient` event for this
+        // consensus state/height was emitted. We expect that these events are received in the exact
+        // same order in which they were emitted.
+        match events.first() {
+            Some(IbcEvent::UpdateClient(event)) => Ok(event.height()),
+            Some(event) => Err(LinkError::unexpected_event(event.clone())),
+            None => Err(LinkError::unexpected_event(IbcEvent::default())),
+        }
+    }
+
+    /// Loops over `tx_events` and returns a tuple of optional events where the first element is a
+    /// `ChainError` variant, the second one is an `UpdateClient` variant and the third one is any
+    /// other variant (usually `ClientMisbehaviour`). This function is essentially just an
+    /// `Iterator::find()` for multiple variants with a single pass.
+    #[inline]
+    fn event_per_type(
+        mut tx_events: Vec<IbcEvent>,
+    ) -> (
+        Option<IbcEvent>,
+        Option<UpdateClientEvent>,
+        Option<IbcEvent>,
+    ) {
+        let mut error = None;
+        let mut update = None;
+        let mut other = None;
+
+        while let Some(event) = tx_events.pop() {
+            match event {
+                IbcEvent::ChainError(_) => error = Some(event),
+                IbcEvent::UpdateClient(event) => update = Some(event),
+                _ => other = Some(event),
+            }
```

Comment:
```suggestion
        Option<ibc::core::ics02_client::events::ClientMisbehaviour>,
    ) {
        let mut error = None;
        let mut update = None;
        let mut other = None;

        while let Some(event) = tx_events.pop() {
            match event {
                IbcEvent::ChainError(_) => error = Some(event),
                IbcEvent::UpdateClient(event) => update = Some(event),
                IbcEvent::ClientMisbehaviour(event) => other = Some(event),
                _ => {}
            }
        }
```

It looks like here we are only interested in the `ClientMisbehaviour` for the `other` field. So probably it is clearer to specify that explicitly.

Otherwise, it is also possible to have events that are not `ClientMisbehaviour` overriding `other` and make that information lost.

On the other hand, I am not very sure why you are specifically matching `None` for `other` at the call site. Are those implying that there were no `ClientMisbehaviour` event, or that there were no other events other than `ChainError`, `UpdateClient`?

Reply:
So the problem was that I wasn't able to find a spec that guarantees that a client update will only emit these 3 types of events and that there will be only one such event per type. And then we're not sure how to handle those other events. Looking at ibc-go, it does seem like only these 3 types of events are possible but then again that is SDK specific code. I like your idea of changing `event_per_type()` to return only `ClientMisbehaviour` events. With that we can then match for the `ChainError` and `ClientMisbehaviour` events first, but we still need to make sure that all update messages succeeded. The good thing is that ibc-go emits `UpdateClient` events even for redundant updates so we will always have one event per update message. And I think it is reasonable to assume that if there are no `ChainError`s and `ClientMisbehaviour`s, then all updates succeeded. WDYT?

Reply:
https://github.com/informalsystems/ibc-rs/pull/1917/commits/14dcb2ae44e47c17a0b03a56aa32d1a74840ee3e

Reply:
TBH I feel that the current approach is pretty ad hoc and unclear at handling the logic, but I also don't know any way of improving it without significant rewriting of the code. So I think we can leave that as is now.

If necessary, we can investigate how the ibc-go really works later on and improve the logic.